### PR TITLE
Fix: add custom setter for NumericAttributesForFiltering

### PR DESCRIPTION
**Pull request by TomKlotzPro**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | yes
| New feature?      | no    
| BC breaks?        | no     
| Related Issue     | 
| Need Doc update   | no

## Describe your change
- Include null for NewtonSoft on `NumericAttributesForFiltering` parameter.
- Add the method [ShouldSerialize](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_JsonProperty_ShouldSerialize.htm) from newtonsoft to avoid serialize empty field.
- Add a dedicated setter for `NumericAttributesForFiltering`. 
- Add tests with `null` for `NumericAttributesForFiltering`

## What problem is this fixing?

Our C# client is using newtonsoft.json to serialize our parameter. This serializer is configurable and for the C# client we exclude null field in our request. However for a specific attribute on settings `NumericAttributesForFiltering` we need to send null to the API.

### CODE REVIEW

File: src/Algolia.Search/Models/Settings/IndexSettings.cs [hunk 1]
```diff
@@ -282,10 +282,32 @@ private List<string> Slaves
 
         // performance
 
+        private bool _numericAttributesForFilteringWasSet = false;
+
+        /// <summary>
+        /// List of numeric attributes that can be used as numerical filters.
```
Comment (Ant-hem) [inline src/Algolia.Search/Models/Settings/IndexSettings.cs]: Please explain why this method is here and link to the documentation: 
https://www.newtonsoft.com/json/help/html/ConditionalProperties.htm

File: src/Algolia.Search.Test/EndToEnd/Index/SettingsTest.cs [hunk 2]
```diff
@@ -184,6 +183,15 @@ public async Task TestSettings()
...
                 spceficPropertiesCheck.ToArray()));
 
+            // Set new value for NumericAttributesForFiltering
+            settings.NumericAttributesForFiltering = null;
+
+            var saveSettingsResponseAfterChangesNumericAttributesForFiltering = await _index.SetSettingsAsync(settings);
+            saveSettingsResponseAfterChangesNumericAttributesForFiltering.Wait();
+
+            var getSettingsResponseAfterChangesNumericAttributesForFiltering = await _index.GetSettingsAsync();
+            Assert.AreEqual(null, getSettingsResponseAfterChangesNumericAttributesForFiltering.NumericAttributesForFiltering);
```
Comment (Ant-hem) [inline src/Algolia.Search.Test/EndToEnd/Index/SettingsTest.cs:193]: 🎉

File: src/Algolia.Search.Test/EndToEnd/Index/SettingsTest.cs [hunk 3]
```diff
@@ -184,6 +183,15 @@ public async Task TestSettings()
...
+
```
Comment (Ant-hem) [inline src/Algolia.Search.Test/EndToEnd/Index/SettingsTest.cs:194]: Please add another line when you are setting `NumericAttributesForFiltering` to a `List<String>`, like that we have the full loop. `string[]` => `null` => `string[]`.

File: src/Algolia.Search/Models/Settings/IndexSettings.cs [hunk 4]
```diff
@@ -282,10 +282,32 @@ private List<string> Slaves
...
+            }
+            set
+            {
+                this.NumericAttributesForFiltering = value;
+                _numericAttributesForFilteringWasSet = true;
+            }
+        }
 
         // Handling legacy index settings
         [JsonProperty("numericAttributesToIndex")]
```
Comment (Ant-hem) [inline src/Algolia.Search/Models/Settings/IndexSettings.cs:313]: We'll have to modify this setter as well (legacy naming) and test it.

**Review by Ant-hem — COMMENTED** [review]
Nice, good job. We're almost good, I have a few remarks.

**Comment by Ant-hem** [comment]
Thanks! We're almost there, let's add a test final test in the file above with a `null` values.

https://github.com/algolia/algoliasearch-client-csharp/blob/bd42c82ae3ed3a4a14959ce0687f2884ae8cdb4f/src/Algolia.Search.Test/Serializer/SerializerTest.cs#L592-L609

File: src/Algolia.Search/Models/Settings/IndexSettings.cs [hunk 5]
```diff
@@ -281,20 +281,44 @@ private List<string> Slaves
...
+        /// <summary>
+        /// ShouldSerialize predicate from NewtonSoft.json to avoid serializing empty field.
+        /// https://www.newtonsoft.com/json/help/html/ConditionalProperties.htm
+        /// </summary>
+        public bool ShouldSerializeNumericAttributesForFiltering()
+        {
+            return _numericAttributesForFilteringWasSet;
+        }
+
+        private List<string> _numericAttributesForFiltering;
```
Comment (Ant-hem) [inline src/Algolia.Search/Models/Settings/IndexSettings.cs:296]: question: shouldn't we add `[JsonIgnore]` here? Can you make sure that `_numericAttributesForFiltering` isn't send in the payload? If yes we're good to merge.

Comment (TomKlotzPro) [reply src/Algolia.Search/Models/Settings/IndexSettings.cs:296]: I've just check and this private field isn't sent in the payload 🥳

**Review by Ant-hem — APPROVED** [review]
